File: corretor.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <cmath>
#include <sstream>

using namespace std;

vector <string> dicionario;
vector <string> palavrasDoTexto;
string texto;

void setup();
void dividirTexto();
bool pertenceAoDicionario(string palavra);
void varrerPalavras();
void corrigirTexto();
void menuPrincipal();
void menuCorrecao(string palavra);
void corrigir(string palavra);
void palavrasParecidas(string palavra);
void adicionar(string palavra);
bool primeiraVerificacao(string palavra, string palavraDoDicionario);
bool segundaVerificacao(string palavra, string palavraDoDicionario);


int main () {
	setup();
	menuPrincipal();
	return (0);
}

void setup() {

	cout << "" << endl
			 << "CORRETOR ORTOGRÁFICO!" << endl
			 << "" << endl
			 << "Primeiramente, iremos configurar o dicionário." << endl
	     << "Digite as palavras que você deseja incluir no dicionário: " << endl
	     << "obs: ao finalizar, digite \"end-dicionario\". " << endl;
	string entrada;
	while (entrada != "end-dicionario") {
		cin >> entrada;
		cin.ignore();
		dicionario.push_back(entrada);
	}
	dicionario.pop_back();
}

vector<string> split(string str, char delimiter) {
  vector<string> internal;
  stringstream ss(str); // Turn the string into a stream.
  string tok;

  while(getline(ss, tok, delimiter)) {
    internal.push_back(tok);
  }

  return internal;
}

void dividirTexto() {

	palavrasDoTexto = split(texto, ' ');


}

bool pertenceAoDicionario(string palavra) {
	for (int i = 0; i < dicionario.size(); i++) {
		if(palavra == dicionario[i]) {
			return true;
		}
	}
	return false;
}

void varrerPalavras() {

	dividirTexto();

	for(int i = 0; i < palavrasDoTexto.size(); i++) {
		if(!pertenceAoDicionario(palavrasDoTexto[i])) {
			menuCorrecao(palavrasDoTexto[i]);
		}
	}

	cout << "" << endl
			 << "=======> Texto após correção: ";
	for (int i = 0; i < palavrasDoTexto.size(); i++) {
		cout << palavrasDoTexto[i] + " ";
	}

	cout << "" << endl;

}

void corrigirTexto() {
	cout << "" << endl
			 << "CORREÇÃO DE TEXTOS" << endl
	     << "" << endl
			 << "Digite o texto que você deseja corrigir: " << endl;
	getline (cin,texto);
	varrerPalavras();
}

void menuPrincipal() {
	int option;

	while(option != 2) {

		cout << "" << endl
	     	<< "MENU PRINCIPAL" << endl
				 << "" << endl
	     	<< "Escolha uma Opção:" << endl
				 << "1 - Corrigir texto" << endl
				 << "2 - Finalizar o programa" << endl;
		cin >> option;
		cin.ignore();


		switch (option) {

			case 1:
				corrigirTexto();
				break;
			case 2:
				return;
				break;
			default:
				cout << "Opção inválida" << endl;
				break;
		}
	}

}


void menuCorrecao(string palavra) {
	int option;
	// O print deve adicionar a palavra errada referente	
	cout << "" << endl
			 << "Palavra errada ----> " << palavra << endl
			 << "1 - CORRIGIR" << endl
			 << "2 - ADICIONAR" << endl
			 << "3 - IGNORAR" << endl << endl
			 << "Escolha opção:" << endl;

			cin >> option;
	 		cin.ignore();

	 		switch (option) {

	 			case 1:
	 				corrigir(palavra);
	 				break;
	 			case 2:
	 				adicionar(palavra);
	 				break;
				case 3:
					break;
	 			default:
	 				cout << "Opção inválida" << endl;
	 				break;

	 		}
}

bool primeiraVerificacao(string palavra, string palavraDoDicionario) {
	int diferencaDeTamanho = fabs(palavra.length() - palavraDoDicionario.length());
	int letrasErradas = 0;
	int palavra1 = palavra.length();
	int palavra2 = palavraDoDicionario.length();
	int count;
	if(diferencaDeTamanho > 2) {
		return false;	
	}

	if (palavra1 > palavra2) {
		count = palavra1;
	} else {
		count = palavra2;
	}

	for (int i = 0; i < count; i++) {
		if(palavra[i] != palavraDoDicionario[i]) {
			letrasErradas++;
		}
	}
	if(letrasErradas > 2) {
		return false;
	} else {
		return true;
	}
}

bool segundaVerificacao(string palavra, string palavraDoDicionario) {
	sort(palavra.begin(), palavra.end());
	sort(palavraDoDicionario.begin(), palavraDoDicionario.end());
	if(palavra == palavraDoDicionario) {
		return true;
	} else {
		return false;
	}

}

void corrigir(string palavra) {
	palavrasParecidas(palavra);
}

void palavrasParecidas(string palavra) {
	vector <string> parecidas;

	for (int i = 0; i < dicionario.size(); i++) {
		if(primeiraVerificacao(palavra, dicionario[i]) || segundaVerificacao(palavra, dicionario[i])) {
			parecidas.push_back(dicionario[i]);
		}
	}

	if (parecidas.size() == 0) {
		cout <<"Não há sugestões. Escolha outra opção."<< endl;
		menuCorrecao(palavra);
		return;
	}

	int index = 1;
	for (int i = 0; i < parecidas.size(); i++) {
		cout << index <<". "<< parecidas[i]<< endl;
		index++;
	}

	int option;

	cin >> option;
	cin.ignore();

	replace(palavrasDoTexto.begin(), palavrasDoTexto.end(), palavra, parecidas[option-1]);


}

void adicionar(string palavra) {
	dicionario.push_back(palavra);
}

File: corretor.hs
import System.IO
import Data.List

menuDicionario :: String -> IO()
menuDicionario palavra
	| palavra == "end-dicionario" = putStr ""
	| otherwise = do
		escreverDicionario palavra
		palavra <- getLine
		menuDicionario palavra

lerDicionario :: IO()
lerDicionario = do
	contents <- readFile "dicionario.txt"
	putStr contents

salvarTexto :: IO()
salvarTexto = do
	texto <- getLine
	writeFile "texto.txt" texto

escreverDicionario :: String -> IO()
escreverDicionario palavra = do
	appendFile "dicionario.txt" palavra
	appendFile "dicionario.txt" " "

corrigirTexto :: IO()
corrigirTexto = do
	putStrLn "\nCORREÇÃO DE TEXTOS\n"
	putStrLn "Digite o texto que você deseja corrigir: "
	salvarTexto
	varrerPalavras

	

--verificar as palavras do texto, no dicionario
varrerPalavras :: IO()
varrerPalavras = do
	palavrasTexto <- readFile "texto.txt"
	palavrasDicionario <- readFile "dicionario.txt"
	let arrayTexto = (words palavrasTexto)
	let arrayDicionario = (words palavrasDicionario)
	verificarCorretude arrayTexto arrayDicionario
	

verificarCorretude :: [String] -> [String] -> IO()
verificarCorretude palavras dicionario
	| palavras == [] = putStrLn "acabou"
	| otherwise = do
		if head palavras `elem` dicionario then
			salvarPalavraCorreta palavras dicionario 
			else 
			corrigirPalavra palavras dicionario

salvarPalavraCorreta :: [String] -> [String] -> IO()
salvarPalavraCorreta palavras dicionario = do
	appendFile "textoCorrigido.txt" (head palavras)
	appendFile "textoCorrigido.txt" " "
	verificarCorretude (tail palavras) dicionario

corrigirPalavra :: [String] -> [String] -> IO()
corrigirPalavra palavras dicionario = do
	let palavra = head palavras
	let opcoes = [x | x <- dicionario, segundaVerificacao x palavra || primeiraVerificacao x palavra]
	menuCorrecao palavra opcoes
	corrigirPalavra (tail palavras) dicionario
	

menuCorrecao :: String -> [String] -> IO()
menuCorrecao palavra opcoes = do
	putStrLn palavra
	putStrLn "Palavra errada ----> " 
	putStrLn "Voce quis dizer "
	putStr(head opcoes)
	putStr "? \n1-Sim\n2-Nao\n" 
	opcao <- getLine
	if(opcao == "1") then adicionarAoTextoFinal (head opcoes)
	else menuCorrecao palavra (tail opcoes)

adicionarAoTextoFinal :: String -> IO()
adicionarAoTextoFinal palavra = do
	appendFile "textoCorrigido.txt" palavra
	appendFile "textoCorrigido.txt" " "

menuPrincipal :: String -> IO()
menuPrincipal opcao
	| opcao == "1" = do
		corrigirTexto
		putStrLn "MENU PRINCIPAL\nEscolha uma Opção:\n1 - Corrigir texto\n2 - Finalizar o programa"
		opcao <- getLine
		menuPrincipal opcao
	| otherwise = putStr("")

primeiraVerificacao :: [Char] -> [Char] -> Bool
primeiraVerificacao x y
	| diferenca <= 2 = True
	| otherwise = False
	where diferenca = diferencaLista x y

diferencaLista :: [Char] -> [Char] -> Int
diferencaLista x y
	| xMaior = (length (x \\ y))
	| otherwise = (length (y \\ x))
	where xMaior = (length x) > (length y)

segundaVerificacao :: String -> String -> Bool
segundaVerificacao x y
	| (sort x == sort y) = True
	| otherwise = False

main = do
	putStrLn "CORRETOR ORTOGRAFICO!\n"
	putStrLn "Primeiramente, iremos configurar o dicionário.\nDigite as palavras que você deseja incluir no dicionário:\nobs: ao finalizar, digite \"end-dicionario\""
	palavra <- getLine
	menuDicionario palavra
	lerDicionario

	putStrLn "\n\nMENU PRINCIPAL\nEscolha uma Opção:\n1 - Corrigir texto\n2 - Finalizar o programa"

	opcao <- getLine
	menuPrincipal opcao
